File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    answer = 0
    n = len(jobs) # 작업 개수
    disk = [] # 디스크 컨트롤러
    now = 0 # 현재 시간
    jobs.sort() # 시작 시간 기준 정렬
    i = 0 # job 번호
    while i<n or disk:
        while i<n and jobs[i][0] <= now: # now 이내 작업 모두 디스크에 저장
            s, l = jobs[i]
            heapq.heappush(disk, (l, s))
            i += 1
        if disk:
            l, s = heapq.heappop(disk)
            now += l
            answer += now-s
        else:
            now = jobs[i][0]
    return answer//n

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    static class Job {
        int start;
        int length;
        Job(int start, int length) {
            this.start = start;
            this.length = length;
        }
    }
    
    public int solution(int[][] jobs) {
        int answer = 0;
        int n = jobs.length;
        int now = 0;
        int i = 0;
        
        Arrays.sort(jobs, (a, b) -> Integer.compare(a[0], b[0]));
        PriorityQueue<Job> disk = 
            new PriorityQueue<>((a, b) -> Integer.compare(a.length, b.length));
        while (i<n || !disk.isEmpty()) {
            while (i<n && jobs[i][0] <= now) {
                disk.offer(new Job(jobs[i][0], jobs[i][1]));
                i++;
            }
        
            if (!disk.isEmpty()) {
                Job job = disk.poll();  
                now += job.length;
                answer += now-job.start;
            } else {
                now = jobs[i][0];
            }
        }
    return answer/n;
    }
}